rfc-160-using-es6-modules-in-govuk_publishing_components.md

```diff
@@ -0,0 +1,428 @@
+# Use ES6 Modules in govuk_publishing_components
+
+## Summary
+
+The approach to authoring JavaScript in [govuk_publishing_components](https://github.com/alphagov/govuk_publishing_components) should change from adopting the [ES5 standard](https://www.w3schools.com/js/js_es5.asp) to instead utilising the modular JavaScript introduced with [ES6](https://www.w3schools.com/js/js_es6.asp). This would mean govuk_publishing_components could use contemporary JavaScript, now supported for several years by the most popular browsers. Eventually, we should move the JavaScript over to a separate npm module. As intermediary step, we will integrate the bundler [Rollup](https://rollupjs.org/) using [jsbundling-rails](https://github.com/rails/jsbundling-rails). This would allow us to deliver separate Javascript files to our frontend applications, enabling [individual component javascript loading](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-149-switch-to-per-page-asset-loading.md). Individual component javascript loading will allows to check browser support of different files as well as provide performance benefits to all of GOV.UK. In addition we can use Rollup to transpile ES6 Javascript to ES5 which could be used to enable analytics features to continue working with Internet Explorer, which has limited support for ES6.
+
+This RFC builds upon ideas from:
+- [RFC #149: Serve component CSS and Javascript as individual files](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-149-switch-to-per-page-asset-loading.md)
+- [This gem is coupled to sprockets/asset pipeline](https://github.com/alphagov/govuk_publishing_components/issues/505)
+
+(Thanks to Kevin for providing lots of insight with Rails, govuk_publishing_components and being very helpful throughout the research process!!)
+
+## Problem
+
+### The future of IE11 support on GOV.UK
+
+To support Internet Explorer 11 (IE11) all Javascript in govuk_publishing_components is written in the ES5 standard. This is because IE11 does not support ES6. However lately in the GOV.UK Frontend Community there have been calls to drop JS support for IE11. As of June 2022, [Microsoft no longer supports IE11](https://learn.microsoft.com/en-us/lifecycle/products/internet-explorer-11). From 24th of April to the 21st of May, the number of users that accessed GOV.UK using IE11 was 26,205. This is 0.067% of all users (38,821,598) who accessed GOV.UK during that time. The concensus in the community is that we should eventually drop support for IE11 but the timeline for this has yet to be established. Part of the complexity is that we would still want analytics to work for the not insigificant number of IE11 users.
```

Comment:
Data point: Between 10th August and 10th September '23 there were 18,841 users of IE, 18,704 of those on IE11. The next major release of the design system will drop js support for IE and is expected before the end of the year. The only pragmatic approach for GOV.UK to take is to fall in line for components inherited from the design system at which point supporting js for IE elsewhere becomes questionable in terms of effort:benefit ratio. The exceptions to this would be cookie acceptance and tracking.